Upsell card reader banner re-appears after dismissal and doing the pull to refresh on the order listing screen.

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #7445 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Upsell card reader banner re-appears after dismissal and doing the pull to refresh on the order listing screen.

**Steps to reproduce**

1. Navigate to the order listing screen
2. Dismiss the banner
3. **Without navigating to any other screen**, Pull to refresh in the order listing screen.
4. Notice the Banner re-appears again.

This PR fixes this as well as refactors the Banner display requirements logic.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/1331230/191910420-44f276f6-c371-4820-bf70-ddc3ec2388f7.mp4



- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Refactor banner display requirements in the order listing screen.

This commit also fixes a bug where the banner after dismissal would reappear when pull to refreshed on the order listing screen.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -456,6 +429,13 @@ class OrderListFragment :
         }
     }
 
+    private fun updateBannerState(orderListSize: Int) {
```
This extra VM -> Fragment -> VM communication purely because you want to know if it's in the landscape? If yes, then probably having some `LandscapeChecker` wrapper class and injecting in directly to the VM will improve the code and testability

kidinov reviewed (approved):
The bug seems to be fixed. I left generic suggesting regarding code improvement. Please feel free to merge that if you want to land it in the next release and then if you think the suggestion is valuable to make an improvement

AnirudhBhat committed:
Refactor the VM -> Fragment -> VM transition to check the banner state.

This commit checks for the banner state in the view model after fetching the orders list.

The banner state is being triggered from the fragment as well to make sure the banner disappears in the Landscape mode.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -199,11 +198,12 @@ class OrderListFragment :
-            lifecycleScope.launch {
-                viewModel.updateBannerState()
-            }
+        updateBannerState()
```
We could probably just use `map` on `_pagedListData` and call `updateBannerState` from there. In this case, we could also make it not suspend and just lunch view model scope coroutine from there

AnirudhBhat replied:
>We could probably just use map on _pagedListData and call updateBannerState from there

The only reason for updating the banner state from the view here is to make sure the banner is not being displayed in landscape mode. The view model doesn't know when the view goes into the landscape mode and someone needs to tell it. 

Another way we can solve this is by checking for the landscape mode in `bannerDisplayViewLogic` method. In this approach, the only downside is that we cannot unit test the logic for banner display in landscape/portrait mode

Let me know if you have another approach in your mind.

kidinov replied:
> The only reason for updating the banner state from the view here is to make sure the banner is not being displayed in landscape mode. The view model doesn't know when the view goes into the landscape mode and someone needs to tell it.

Please correct me if I am wrong, but on configuration change, the view model with its livedatas will push the latest state to the view, that's why `map` will be called and `updateBannerState` as well, where you will change already if it's landscape or not

AnirudhBhat replied:
I think `map` gets called when the `livedata` it's observing is set some value. When a configuration change happens, the latest result from the livedatas are simply pushed into its observers, and `map` itself doesn't get called. WDYT?

kidinov replied:
> the latest result from the livedatas are simply pushed into its observers, and map itself doesn't get called. WDYT?

That would be really magical behavior that somehow the code inside of map would be ignored in certain situations.  I assume you tried to map it on `_pagedListData`?

I didn't find a good approach to keep mutable mapped life data hidden and with a nice syntax

One might be (not really "kotlinish" way):

```
    private val _pagedListData = MediatorLiveData<PagedOrdersList>()
    fun getPagedListData(): LiveData<PagedOrdersList> = _pagedListData.map {
            println("pagedListData map")
            it
        }
```

Not ideal, but probably a bit better than calling pushing state multiple times to the composable

Having said that, I don't have a strong opinion on this, so it's totally up to you. You might find know a better approach here

AnirudhBhat replied:
You are right. I had this implementation initially
```
private val _pagedListData = MediatorLiveData<PagedOrdersList>()
val pagedListData: LiveData<PagedOrdersList> = _pagedListData.map {
            println("pagedListData map")
            it
        }
```
It never occurred to me that the `val` caches its value 🤦 I have added a custom getter for the `val` and all of the banner state handling is being done in the `ViewModel` now. 

Thanks for the `map` idea 💡 

I'll be merging this PR now since I need to get this in by tomorrow. Please feel free to drop any other suggestions and I will make sure to look into those for the next release.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -188,11 +191,21 @@ class OrderListViewModel @Inject constructor(
+        updateBannerState(landscapeChecker, _pagedListData.value.isNullOrEmpty())
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
```
Probably pushing something to the `pagedListWrapper` you could avoid having it public?

kidinov reviewed (commented):
Thank you for addressing my comments!  Sorry, but I left a couple of more =)

AnirudhBhat committed:
Update the banner state from the view model. View doesn't need to handle anything related to the banner apart from displaying it.
